### ISSUE TITLE
Non-cached OfficeRuntime.auth.getAccessToken is ~500-600ms slower on New Outlook and throws 13008 errors vs. Classic and Web Outlooks

### ISSUE THREAD

**Issue by pepso** [OfficeDev/office-js#5112]
## Your Environment
* Platform: PC desktop
* Host: Outlook
* Office version number: New Outlook
* Operating System: Windows 10 / 11
 
## Expected behavior
The New Outlook's getAccessToken should return broadly speaking as New and Classic Outlooks because this is "just" a Azure AD v2 token wrapper with caching layer. 

Also we expect the New Outlook not to throw 13008 errors when there is no add-in SSO token fetch calls in progress.

## Current behavior
New outlook takes ~2s (~500-600ms longer than other Outlook types) to return refreshed token and sometimes throws 13008 errors when there is no parallel token calls ongoing.

## Steps to reproduce
- Have a SSO enabled Add-in
- Configure the SSO to use obo flow
- Skip the cached token calls
- Use option: `{ forMSGraphAccess: true }`
- Measure the time it takes to fetch the token in different Outlook types

# Provide additional details
1) We use OBO flow so we don't have data if this is impacting other flows
2) New outlook refreshes the token ~40% slower than other types based on 10'000 calls we looked at

## Context
User is greeted by 5s warnings too often in New Outlook when 40% of the available time on average is spent on fetching the token. This is compounded by the network layer variability and including this results to 1.6% of New Outlook SSO token fetch calls taking over 5 seconds to return while this is 0.3% for Classic Outlook.

The 500ms makes a big difference to create a consistent and good user experience on the New Outlook for our users because we need to still use this token on the server side (with token exchange and API calls taking 1+ seconds).

Think about the current flow in context of 5 second warning in New Outlook:
- Add-in sandbox runtime available from LaunchEvent triggered: 0-500ms
- Call to getAccessToken(): 2 seconds
- External API call: >1 seconds (obo exchange 500 ms and using the token: 500 ms).

Basically the OBO flow on average already ate 3-3.5 seconds of the available 5 seconds warning buffer which leaves us 1.5 seconds to handle all the other add-in logic like appointment field fetching, notifications etc.

## Useful logs
10'000 data samples we have points to these averages:
Classic Outlook: 1398 ms
Outlook Web App: 1320 ms
New Outlook Windows: 1934 ms

The timing avg data is consistent when excluding the limited weekend data (it's been like this for months):
![image](https://github.com/user-attachments/assets/d720c2f8-4356-4285-afb2-496a37dc88fb)

Thank you for taking the time to report an issue. Our triage team will respond to you in less than 72 hours. Normally, response time is <10 hours Monday through Friday. We do not triage on weekends.

**Comment by mobisw-msft** [comment]
Hey @pepso ,

Unfortunately, again- we are not able to see this in our telemetry right away. But we are investigating this issue further.

**Comment by rkpathak** [comment]
We are monitoring the telemetry for getAccessToken API to compare the performance issue. 
Regarding 13008 error, we see that most of the time this error is observed in case of auto run/launch event use cases. Could you please share a sample addin which can reproduce this problem?

**Comment by pepso** [comment]
Hi @rkpathak, I added you as a collaborator to the repo "ms-comms" where we have info about the tenant and add-in ID, and link to current manifest.xml (please see file: issue-4917.

Our Add-in is a launch event based and in the Office.initialize we rely on cached configuration (non-expired) or load new from our servers using obo/getAccountToken flow.

During development:
We saw the 13008 surprisingly often in our logs but sometimes it is actually related to double calling the API between hot reloads which we can attribute to the development flow reasons.

When it comes to the production add-in:
Chance for users to open 2 appointments at the same time => resulting into 2 separate sandboxes => and they do this when there's no cached configuration available (expires in 24 hours) => Calling getAccountToken actually twice from 2 separate sandboxes

We assess above as an unlikely scenario in user's normal calendar usage pattern. 

To "manage" these long running calls and errors we have resulted to following safe guards:
1) Add a delayed retry in case of an exception
2) Non-classic Outlook notification for long running process for UX purposes
3) Parallel call protection solution within a sandbox

Our 13008 error rate in production is something like ~1% of users per day which we recover every time using the delayed retry.

**Comment by rkpathak** [comment]
The API could be getting called from other installed addins as well. Delayed retry is a good safeguard for this. If your questions are answered please close the issue.

**Comment by pepso** [comment]
Hi @rkpathak 

This is a bug report / performance issue about 500ms longer API call on New Outlook vs. Web and Classic Outlooks and there is no answer to reasons why this is the case or when it would be fixed. 

The 13008 errors are mentioned as a way to identify the culprit. Update to the [docs](https://learn.microsoft.com/en-us/office/dev/add-ins/develop/troubleshoot-sso-in-office-add-ins#13008) might be good idea to include New Outlook where this error can arise from if it is indeed expected.

**Comment by arpitsangwanmsft** [comment]
Hey @pepso!
1) For the 13008 errors, can you please clarify if you observe those in unexpected scenarios as well, as you have called out these are expected if the request to get token is already in queue and another request is made. If yes, then can you please mention repro scenario/steps.

2) We could not observe the trend of New Outlook being slower than Web client. In fact, getAccessToken is faster for New-Outlook as per our telemetry in general. To further investigate telemetry for your particular tenant, can you please `enable troubleshooting logging for 24 hours` for a bunch of users so that we can get unsampled data: 

![image](https://github.com/user-attachments/assets/6fe4f167-834e-4258-a1fa-5972ac99199c)

3) Even though, we could not observe New-Outlook being slower than web, we could observe getAccessToken taking over 2 seconds for few cases, for which we have created an issue in our backlog - tracking id-323467

**Comment by pepso** [comment]
Hi @arpitsangwanmsft,

### 1) 13008 exceptions when calling getAccountToken 
We see these exceptions when our add-in is NOT calling the getAccountToken in parallel.

I understood from @rkpathak that these can be caused by 3rd party Add-in "occupying" the getAccountToken end point and hence we would see these errors. We find this a bit unexpected design decision if this is the case.

**We do not expect:**
Seeing 13008 errors from the Outlook API using our unique WebApplicationInfo/API Resource when completely unrelated Add-in is using their own WebAppInfo/getAccountToken end point. 

Our pseudo steps in the production which results to ~1% of 13008 error rate for the getAccountToken calls we do in this flow (New Outlook only):
1) Office.initialize => wait or set localStorage flag init ongoing true (also ensures the flag is owned by this "thread")
2) Office.initialize => call getAccountToken and call our AWS hosted API to fetch a config (using the OBO token exchange flow) to add-in global variable and localStorage (for 24 hours)
3) Office.initialize => set localStorage flag init ongoing false
4) All launch events => get config from global variable => Classic Outlook can call the init procedure from here because there the V8 sandbox does not support/ensure Office.initialize before executing the launch events. 

Only way we can think of calls from our add-in happening in parallel is that there is 2 completely separate sandboxes which access separate localStorage (via Office APIs) and hence the protection method we have in place is not able to ensure we call the api one at a time.

**What is expected:**
In our development flow we completely get there are parallel calls taking place to same getAccount token endpoint (side-load/dev/UAT/production), and these do make sense as they happen against the same WebApplication/API Resource configuration. Our Add-in's parallel call protector cannot access the same localStorage as it's not shared between the Add-in versions (unique IDs).

### 2) Tenant and their call time experience
We see a lot of call time fluctuation between tenants and their geographical distribution in the end point. 

Here's a getAccountToken calls for a tenant of our Add-in users whose are located in 1 country and 2 different cities and use the add-in ~85% from Classic, ~15% New Outlook, and ~1% Web and Mac Outlooks:
![image](https://github.com/user-attachments/assets/2eccad2a-5b5b-401e-9f02-2b9b1898817e)

Here's the averages of those calls:
![image](https://github.com/user-attachments/assets/4aeab473-bc95-4f1e-836b-f7a3ad8ae5e4)

As you can see, we are in a totally different ball park compared to "few calls taking over 2 seconds"! 

When we try to replicate this in our own internal cloud infrastructure we barely see getAccountToken calls times over 1 second, and this client records calls taking over 5 seconds every day. We have excluded VPN's, poor network reception, etc. many times over. In the same connected environment the fetch call to our AWS hosted API with the token (and OBO exchange + MS Graph Call) takes on average ~1 seconds.

The tenant in question have had the extended 24 hour logging turned on last week on multiple days while we have been investigating the performance issues of #4917 and #5040. I'm going to add you to the same private repo where you get the tenant ID and add-in ID, and the manifest.xml so you can query to correct/unsampled telemetry to investigate the getAccountToken calls based on the last weeks extended telemetry. You will find the information from file named "issue-4917". Our users call the getAccountToken rarely, approximately once a day due to the level of caching we have in place to overcome the performance issues we see in New Outlook.

### 3) Over 2 second call times
Out of curiosity, are you from the Azure v2 end point team, or from the New Outlook's side? I'm asking this to understand if the 2 second limit is expected to cover the whole journey from Add-in -> Office.JS -> Azure Token API end point, or are we talking about Office.JS -> Azure Token API end point response times?

**Comment by pepso** [comment]
Hi @arpitsangwanmsft,

I just spoke with the client and can confirm there's been few users this week who have enabled the 24 debugging on multiple days.

**Comment by arpitsangwanmsft** [comment]
Hey @pepso !
'enable troubleshooting logging for 24 hours' debugging option needs to be re-enabled after 24 hours and please add query parameter: `track=*` to the url when using Outlook on Web for the enabled logs sessions.

That being said, while we are trying to investigate the issue internally can you please rule out:
Some environment or configuration problems that might be contributing to the delays since we are not observing any recent degradation in our general telemetry and this is being observed for a single tenant, for example:

- A proxy server or VPN in the setup might be introducing delays.
- Internal DNS resolver not being correctly configured.
- Firewalls adding delays in processing requests.
- Some browser extension -ad blockers/privacy tools.

Can you confirm if this is a single tenant problem ?

**Comment by arpitsangwanmsft** [comment]
And as for general delay in getAccessToken (taking over 2 seconds), We have around ~13% of the calls taking more than 2 seconds(for web) and around ~9% for New-Outlook, for which we have logged a backlog item as mentioned above. The delay includes `Add-in-> Office.JS -> Azure Token API and back` time.

**Comment by pepso** [comment]
Hi @arpitsangwanmsft,

1) Yes, we have gone through all the common suspects (i.e. proxy/VPN/Firewalls) 
2) Yes, the client added the checkbox on multiple days to expand the data as you suggest
3) Other clients using the add-in are 99.9% on Classic Outlook, so there is no coherent New Outlook data available :(

When we look at the Classic Outlook data for our US clients, we see token fetch average being higher than EU clients and we explain this by assuming our app id is somehow hosted in Europe side vs. client Outlooks residing on US side.

Few questions to narrow this down:
Is our Outlook's getAccountToken spread/avg something you see also in your telemetry? Thinking about the network layer in general, we most certainly should see the same "slowness" on Classic Outlook and the New Outlook if this is the reason for this tenant? 

Makes me also wonder if there is some New Outlook only configuration for network proxies we might have missed?

**Comment by pepso** [comment]
Looking at our own API fetch times New Outlook performs actually better than Outlook so that should exclude any network or proxy issue with New vs. Classic Outlook token fetch.

![image](https://github.com/user-attachments/assets/33f90024-b3a8-456d-ae72-977975235d17)

**Comment by arpitsangwanmsft** [comment]
Hey @pepso!
Yeah, the spread in this latest graph pointed is somewhat in line with our telemetry. But the earlier graph shared which I am assuming is for that particular tenant is not in-line with observed latencies in general. It might be because of intermittent server issues during that week for that particular tenant.
If you observe same delays for other tenants again, keep us posted.

**Comment by pepso** [comment]
@arpitsangwanmsft the chart is our AWS API fetch, not the Token end point. The token end point continues to be mired with delays for New Outlook only. 

Here they are side by side (to exclude the issue is something to do with networking):
![image](https://github.com/user-attachments/assets/34d7ebe1-5b1d-4eb9-b862-86f17869423a)

Only slow down we see is New Outlook getAccountToken. This is not the case for web, nor for the Classic Outlooks. Networking in general is solid for New/Classic/Web as per our own API calls demonstrates.

The original question stands: where is this 500-600ms New Outlook's getAccountToken slow down coming from?

**Comment by rajjha-msft** [comment]
Hey @pepso 

Thanks for reporting this issue regarding the performance of OfficeRuntime.auth.getAccessToken Api. It has been put on our backlog for further investigation. We unfortunately have no timelines to share at this point 

Internal tracking id: Office: 5097217

**Comment by pepso** [comment]
@arpitsangwanmsft and @rajjha-msft since finding out the replicable performance issue in appointment field API and notification API I wanted to ask a question/provide plausive explanation to the slowdown:

When the add-in is running a cold start and does its first call to external API via Office.JS, what's the boot time of service worker in main New Outlook window the add-in needs to wait?

getAccountToken is the first API call (when we do full refresh of the config) which relies on exchange server/external service. If the service worker is not warmed up at the time the add-in is, this could explain why there's the delay in New Outlook only.

**Comment by pepso** [comment]
@rajjha-msft any updates on these slow token fetch API calls?
